docs/design/cuda_graphs.md

```diff
@@ -0,0 +1,226 @@
+
+So we seek a more fine-grained CUDA Graphs solution that can:
+
+* Be explicitly aware of whether a batch is prefill/uniform decode/mixed and should capture/replay the CUDA Graphs accordingly. It should do this because a unified full CUDA Graph for different cases of the same batchsize is usually infeasible (e.g., for many attention backends).
+* Capture full CUDA Graphs while maintaining the ability to use piecewise CUDA Graphs. i.e, can dispatch CUDA Graph-incompatible routines (e.g., cascade attention or mixed prefill/decode batches for some attention backends) into piecewise CUDA Graphs.
+* Achieve centralized control of the CUDA Graphs behavior via a dispatcher, which makes CUDA Graphs dispatching easier to understand and more extensible.
+* Add CUDA Graphs support to models that do not fit vllm's torch.compile integration system design in v1.
+
+We also found that when a batch cannot hit a full CUDA Graph, the host-side eager execution of the flattened compiled FX graph (previous behavior) can be slower than the piecewise compiled FX graph in Python (see <gh-pr:20059>). So, we prefer maintaining the piecewise compilation when enabling full CUDA Graphs to reduce host-side overhead. We can safely do this as full CUDA Graphs and compilation are actually orthogonal to each other.
+
+## `CudagraphModes`
+
+[CUDAGraphMode][vllm.config.compilation.CUDAGraphMode] is the single knob you tune in `CompilationConfig.cudagraph_mode`:
+
+* `NONE` — turn CUDA Graphs off. Good for debugging.
+* `PIECEWISE` — default in v1. It is the most flexible: attention or other CUDA Graphs-incompatible operations stay eager, everything else goes into CUDA Graphs.
+* `FULL` — a single-mode strategy, which only captures full CUDA Graphs for non-uniform batches, then uniform-decode batches reuse the CUDA Graph of non-uniform batch of the same batch_size, since they are compatible; can be good for small models or workloads with small prompts.
+* `FULL_DECODE_ONLY` — full CUDA Graph for uniform decode, eager run for prefill/mixed etc; suitable for decode instances in a P/D setup where prefill is not as important, so we can save some memory.
+* `FULL_AND_PIECEWISE` — full CUDA Graph for uniform decode, piecewise CUDA Graphs for others; generally the most performant setting for most models.
+
+Defaults: If you’re on v1 with piecewise compilation, we default to `PIECEWISE` for safety reasons (for mamba mixer models, it's `FULL_AND_PIECEWISE`). Otherwise, we default to `NONE`.
+
+!!! note
+    We also fuse the subset modes `NONE`, `PIECEWISE`, and `FULL` as the concrete runtime modes for CUDA Graphs dispatching, so they are treated as one of the "decode_mode" or "mixed_mode" at runtime.
```

Comment:
I don't really understand this part. What are subset modes? Is this saying the same thing as the paragraph below?